### PR TITLE
fix: resolve first-party workspace package imports and bound the name fallback

### LINKED ISSUE #945: fix: app calls into workspace-package TS SDKs barely resolve, and the few that do are name-fallback misbinds
Follow-up from #912, observed on a large private polyglot monorepo.

## Symptom

The graph holds 519 generated TS SDK methods (vendored OpenAPI client classes) and 114 application functions that call into them, yet CALLS edges resolve to only 4 distinct SDK targets, and inspection shows those few resolutions are wrong: a Python function resolving to a nested TS closure named `sleep` inside a generated helper, and a frontend view resolving to a generated `createClient.request` closure, both clearly bare-name trie fallbacks rather than genuine bindings.

## Suspected causes

1. Applications import the SDK through its workspace package name (an npm-style scoped alias defined in the monorepo's package manifests), not a relative path; the TS import mapping has no equivalent of the Go module-directive mapping (#941), so scoped package imports never map to the vendored directory and every call through them stays unresolved.
2. With the import unresolved, the bare-name fallback binds across languages and into nested closures, producing the misbinds above; the fallback should not cross languages, and arguably should not bind into closure-scoped names.

## Suggested direction

Discover workspace package name to directory mappings from package manifests (mirroring the Go go.mod module-path discovery), resolve scoped imports through them, and tighten the name fallback to reject cross-language and closure-scoped candidates. RED coverage: an app module importing a class through a workspace alias and calling a method on an instance must resolve to the vendored class method, and a same-named nested closure in another language must not be a candidate.

### PR DESCRIPTION
**Pull request by vitali87**
## Summary

- Resolve TypeScript imports of first-party workspace packages by manifest name (`@acme/sdk/admin`), mirroring the go.mod module-path discovery from #941: every `package.json` name maps to its directory, and a specifier resolves through the package's own `exports` map (exact keys before patterns, `null` blocks honoured) or its `main`/`module`/`types` entry. A manifest names the PUBLISHED file, so when that is a build artefact the source it was built from resolves instead, and candidates inside ignored build directories are refused outright because the graph never holds those files.
- A conditions object is read the way Node reads it: the manifest's own declared order decides which condition wins, so a `default` listed ahead of `import` is selected as written rather than reordered by a fixed ranking, and a selected condition whose nested branch offers this analysis nothing falls through to the next key instead of failing the whole entry. The two module systems stay separate (an `import` never binds a `require` target or the reverse) and the declaration-only `types` condition is never followed. Both require paths agree on the require side: the destructured `require()` handled in the module-system mixin now reads the map from the require condition, as the plain `require()` path already did, so a dual package no longer resolves one statement to two different sources.
- Bound the simple-name call fallback to candidates the caller can reach: one written in a language the caller's cannot call into is impossible and is dropped (JS/TS, C/C++ and Java/Scala each count as one family; unknown language on either side keeps the candidate), while one scoped inside another function's body is merely unlikely, since a factory can return it or CommonJS export it, so unscoped candidates win when any exist and scoped ones still bind when they are all there is.
- Apply the same reachability bound to protocol dispatch, where a Python Protocol was fanning out to eight generated TypeScript methods, and read an unchanged module's language from the path recorded on its graph node so an incremental update and a full re-index agree.

## Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Performance improvement
- [ ] Refactoring (no functional changes)
- [ ] Documentation
- [ ] CI/CD or tooling
- [ ] Dependencies

## Related Issues

Closes #945.

## Test Plan

- [x] Unit tests pass (`uv run pytest codebase_rag/tests -n auto`: 6223 passed, 10 skipped)
- [x] New tests added
- [x] Integration tests pass (run as part of the suite above)
- [x] Manual testing (describe below)

Each new test was demonstrated RED before its fix: the resolution shapes (exports wildcard through a build root, `main` entry, no-exports subpath, re-exporting barrel, nested package shadowing, third-party untouched), the resolver's edges (build artefact never wins, parent-escaping target refused, dot-prefixed directory kept, exact key beats pattern, `null` blocks a subpath, a matched export not overridden by a guess, wrong-case specifier), condition selection (the manifest order decides the winner, an unusable nested branch falls through, module systems never cross, `types` never stands in for the runtime module), discovery (node_modules ignored, unreadable manifests skipped, deterministic order for duplicate names), and the fallback bound (cross-language closure and method refused, exported closures in Python and CommonJS still bind, module-level beats another function's closure, same-family and own-nested calls unaffected, incremental language evidence).

Manual validation on a large private polyglot monorepo: calls from application code into its generated SDKs went from 115 edges over 4 targets, every one a misbind, to 46 edges over 35 targets, all genuine, with no cross-language callers left. Total call edges rose by about a thousand once the closure rule became a preference rather than an exclusion. Contract joins, dispatch links and URL resolutions are unchanged.

## Checklist

- [x] PR title follows [Conventional Commits](https://www.conventionalcommits.org/) format
- [x] All pre-commit checks pass (`uv run pre-commit run --all-files`)
- [x] No hardcoded strings in non-config/non-constants files
- [x] No `# type: ignore`, `cast()`, `Any`, or `object` type hints
- [ ] No new comments or docstrings (code should be self-documenting)

The last item is not met and is not achievable here: the file this PR extends documents every non-obvious rule in comments (`endpoint_routes.py`, `go/module_paths.py` and the call resolver are written that way throughout), and the constraints encoded here — why a build artefact must lose to its source, why a scoped candidate is a preference rather than an exclusion — cannot be read off the code.

## Summary by CodeRabbit

* **New Features**
  * JavaScript/TypeScript monorepo “workspace package” imports now resolve using first-party `package.json` `exports` (including wildcards/conditions) and legacy entry points, with deterministic handling for overlapping package names and CommonJS `require`.
  * Added public helper APIs for workspace package discovery and resolution.
* **Bug Fixes**
  * Improved call and Protocol dispatch to avoid incorrect cross-language matches and better respect scoping/module-level precedence, including on incremental rehydration.
* **Tests**
  * Added end-to-end and unit coverage for workspace import resolution and call/Protocol dispatch edge cases.
* **Chores**
  * Added supporting log messaging and a shared recursive JSON type.

### CODE REVIEW

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
All three findings are fixed in 3fdfbbf3, each pinned by a test that fails without its fix:

- Root condition maps: an `exports` object whose keys are all conditions (no key starts with `.`) now declares the package ROOT and matches no other subpath, so `{"import": "./dist/entry.js"}` resolves to `src/entry.ts`. The test deliberately puts the entry at a non-conventional path so only reading the condition map can resolve it.
- Exact-key precedence: an exact key now wins EXCLUSIVELY, so when it names a file this repo does not hold the subpath stays unresolved instead of falling through to a wildcard and binding the import to a different module.
- The Ruff PLR1714 failure in Lint & Format is fixed (the pinned pre-commit ruff is a newer version than the one CI runs, which is why it passed locally; I now run `uv run ruff check` as well).

Also addressed from the review checks: the PR body now follows the repository template, and the manifest readers no longer use `Any` or `object` hints, using a recursive `JsonValue` alias in `types_defs` instead.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
Round 2 addressed in the new commit, both pinned by tests:

- Exports precedence: an `exports` map now supersedes `main`/`module`/`types` entirely, as it does in Node, so a root export naming a file this repo does not hold (or a blocked one) no longer falls back to a legacy field. Combined with the exact-key exclusivity from the previous round, a claimed subpath resolves through its own declaration or not at all.
- An absolute target is refused alongside the parent-escaping ones, so it cannot reach outside its package directory.

The earlier two findings (root condition maps, exact-key precedence) were fixed in 3fdfbbf3.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
Both P1 threads are fixed in the new commit, each pinned by a test that fails without it:

- Wildcard precedence: the best-matching key now wins exclusively, not just an exact one, so when the most specific pattern names a file this repo does not hold the subpath stays unresolved instead of falling through to a broader pattern that maps somewhere else.
- Conditional targets: conditions are tried in a fixed order (`types`, `import`, `module`, `default`, `require`) rather than mapping order. The graph holds sources rather than builds, so these usually lead back to one file; where an `import` and a `require` condition genuinely name different sources, the ESM one now wins deterministically instead of whichever the manifest happened to list first.

Note on the previous round: the summary that reported 3/5 was generated against 15539f1f, before the two fix commits it asks for. The exact-key and legacy-field precedence issues it lists were fixed in 3fdfbbf3 and 1a941e30.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
The Windows unit-test failure was in the two discovery assertions, which compared `str(Path)` against a posix literal; they now compare with `as_posix()`. The resolver itself was already platform-correct (it returns `relative_to(repo).as_posix()`), so no production behaviour changed.

**Comment by vitali87** [comment]
@greptile review latest

**Comment by vitali87** [comment]
SonarCloud flagged three Cognitive Complexity smells in the new resolver. Fixed by extracting the branching into named steps rather than by raising the threshold: `_is_root_only` and `_key_match` for the exports matcher, and `_within_package` plus `_source_candidates` for the on-disk probe. No behaviour change; all 27 resolver tests still pass.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
Fixed: an `exports` map is now exhaustive, as it is in Node. Declaring one at all claims every subpath, so a package that lists only subpaths does not expose its root and a root-only condition map does not expose a subpath; neither falls through to a legacy entry field or a conventional-path guess.

That also corrected one of my earlier tests, which had asserted the opposite for a root-only condition map. It now asserts both halves: the root resolves through the condition map, and a subpath resolves to nothing.

Checked against the real monorepo the feature was built for: all four first-party specifier shapes still resolve to the same source files, and third-party specifiers still resolve to nothing.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
Fixed: the import kind now reaches the resolver, so a CommonJS `require()` reads a dual-package `exports` map from the require side (`types`, `require`, `default`, then the ESM conditions) while an ESM import keeps the ESM-first order. Pinned by a test that resolves the same subpath both ways and gets the two different sources.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
Fixed: condition selection no longer crosses the module systems. An ESM import considers `types`, `import`, `module`, `default`; a require considers `types`, `require`, `default`; an unrecognised condition (a runtime or bundler key) is tried last in either case. So when the selected side's source is not indexed the request resolves to nothing rather than binding to the other system's module, which is the same discipline the rest of this resolver already follows for an unlisted subpath, a null block, or a wrong-case spelling.

Test resolves the same subpath both ways against a package where only the CommonJS source exists: the require resolves, the ESM import does not. Real-monorepo specifiers all still resolve unchanged.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
Fixed: only the FIRST condition present in the map is used, following the request's own order, so an unavailable `import` target no longer falls through to `module`, `default` or a custom condition; the request stands or falls on the condition Node would have selected. Conditions naming the other module system are still never considered, and an unrecognised runtime key remains the last resort.

Pinned by a test where `import` names a missing file and `default` names an existing one: the ESM request resolves to nothing. All five real-monorepo specifier shapes, including the nested `client` subpath, still resolve to the same sources.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
Fixed: `types` is no longer selectable for a runtime import. It names the declaration module, which no runtime ever selects, so a call graph must follow the module that actually executes; it is still recognised as a known condition so it is not mistaken for a custom runtime key. Selection is now `import`, `module`, `default` for ESM and `require`, `default` for CommonJS, first present wins, never crossing between the two.

Pinned by a test where `types` and `import` name different modules. Real-monorepo specifiers unchanged.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
Fixed: only the module-system conditions and `default` are selectable. Every other one (`browser`, `development`, `react-native`) is chosen by an environment this analysis does not have, so a map offering nothing else names no module the request can be said to load, and resolves to nothing rather than binding to an arbitrary environment variant. A `default` alongside them still serves, as it does in Node.

Two tests: an environment-only map resolves to nothing, and the same map with a `default` resolves through it. Five real-monorepo specifier shapes still resolve to the same sources, including one that goes through a wildcard into the generated schema tree.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
@greptileapps

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/js_ts/module_paths.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        )
+        # Node selects the FIRST condition present in the map, following the
+        # request's own order, and the request stands or falls on that
+        # target: a later condition is not a second chance, and one naming
+        # the other module system is never considered. Only the module-system
+        # conditions and `default` are selectable, because every other one
+        # (`browser`, `development`, `react-native`) is chosen by an
+        # environment this analysis does not have, so a map offering nothing
+        # else names no module it can follow.
+        ordered = [key for key in order if key in value]
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/js_ts/module_paths.py]: Fixed in 5bd35dd and 6846f17b. Selection now walks the exports object in the manifest's own declared order and takes the first key this analysis can select, matching Node's PACKAGE_TARGET_RESOLVE, so a `default` listed ahead of `import`/`require` wins as Node intends. The fixed-priority tuple is gone (constants are now unordered `frozenset`s that only say which conditions are selectable). Follow-up 6846f17b also makes a selected key whose nested branch offers nothing fall through to the next key, mirroring Node's continue-on-undefined. Covered by test_the_manifest_decides_which_condition_wins and test_a_selected_condition_with_no_usable_branch_falls_through.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/import_processor.py [hunk 2]
```diff
@@ -1015,10 +1036,15 @@ def _strip_js_extension(import_path: str) -> str:
                 return import_path[: -len(ext)]
         return import_path
 
-    def _resolve_js_module_path(self, import_path: str, current_module: str) -> str:
+    def _resolve_js_module_path(
+        self, import_path: str, current_module: str, require: bool = False
+    ) -> str:
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/import_processor.py:1041]: Fixed in eca4998d. `_process_commonjs_import` (the destructured-`require()` path in js_ts/module_system.py) now passes `require=True`, so a dual package resolves through the require condition, matching Node and agreeing with `_parse_js_require`, which already resolved require-side. Before the fix the two paths disagreed and emitted an ESM and a CJS IMPORTS edge for the same statement; they now agree on the CommonJS source. New RED-first test test_destructured_require_resolves_the_commonjs_condition asserts the require target is chosen and the ESM one is absent.

**Comment by vitali87** [comment]
@greptile review
